node/core/candidate-validation/src/lib.rs

```diff
@@ -657,6 +654,27 @@ impl ValidationBackend for ValidationHost {
 	}
 }
 
+// Tries executing a PVF a single time (no retries).
+async fn execute_pvf_once(
```

Comment:
I propose to keep this code in `validate_candidate` (don't touch it basically) and introduce a new method to `ValidationBackend` trait
```rust
async fn validate_candidate_with_retry(
    ..,
    can_retry: impl Fn(ValidationError) -> bool
)
```
And give it a default implementation (the one you already have) which calls `validate_candidate`

This would make a code easily testable

Reply:
Done, except without the `can_retry` parameter as I didn't see a need for it. It would make the code more general but isn't it unnecessary abstraction?

Reply:
Whatever you prefer, it was an insignificant suggestion :)